A65-xds-mtls-creds-in-bootstrap.md

```diff
@@ -0,0 +1,97 @@
+A65: mTLS Credentials in xDS Bootstrap File
+----
+* Author(s): @markdroth
+* Approver: @ejona86
+* Status: {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2023-05-24
+* Discussion at: https://groups.google.com/g/grpc-io/c/h_LQnTimyt4
+
+## Abstract
+
+This proposal adds support for configuring the use of mTLS for
+communicating with the xDS server via the xDS bootstrap file.
+
+## Background
+
+[gRFC A27][A27] defines the xDS bootstrap file format for setting
+the channel creds used to communicate with the xDS server, with the
+initial options of `google_default` and `insecure`.  It suggested that a
+general-purpose mechanism could be added to configure arbitrary channel
+creds types, which has subsequently been done, albeit not intended as
+a public API.
+
+[gRFC A29][A29] describes the certificate provider framework used for
+configuring mTLS for data plane communication in xDS.  It also defines
+the file-watcher certificate provider.
+
+This proposal defines a new channel creds type to be used in the
+bootstrap file for using mTLS with the xDS server, leveraging the
+functionality of the file-watcher certificate provider.
+
+### Related Proposals: 
+* [gRFC A27: xDS-Based Global Load Balancing][A27]
+* [gRFC A29: xDS-Based Security for gRPC Clients and Servers][A29]
+
+## Proposal
+
+We will define a new credential type in the bootstrap file called `tls`.
+Its configuration will be essentially the same as that of the file-watcher
+certificate provider described in [gRFC A29][A29].  Specifically, the
+config will look like this:
+
+```json
+{
+  // Path to identity certificate file.
+  // If set, mTLS will be used; if unset, normal TLS will be used.
+  "certificate_file": <string>,
+
+  // Paths to private key and CA certificate files.
+  // If either of these fields are set, both must be set.
+  // If neither is set, system-wide root certs are used.
+  "private_key_file": <string>,
+  "ca_certificate_file": <string>,
+
+  // How often to re-read the certificate files.
+  // Value is the JSON format described for a google.protobuf.Duration
+  // message in https://protobuf.dev/programming-guides/proto3/#json.
+  // If unset, defaults to 10 minutes.
+  "refresh_interval": <string>
+}
+```
+
+The only difference between the file-watcher certificate provider config
+and this one is that in the file-watcher certificate provider, at least
+one of the "certificate_file" or "ca_certificate_file" fields must be
+specified, whereas in this configuration, it is acceptable to specify
+neither one.
+
+Implementations should be able to internally configure the use of the
+file-watcher certificate provider for the certificate-reloading
+functionality.
+
+### Temporary environment variable protection
+
+This feature is not enabled via I/O, so no env var protection is needed.
```

Comment:
It is enabled via I/O, because we read the bootstrap file, and newer bootstrap generators may want to support older clients.

I'd have to dig in to see if that distinction matters in this case though. I know we don't support a list of credentials here, so old clients would appear to fail anyway. There could be complications if we supported multiple servers in the list, but we don't do that either. So I could agree it isn't needed here (especially since this isn't a difficult feature and we're probably not going to have interop tests for it), but the reasoning would need to change.

Reply:
Good point -- what I really meant here was that it's not enabled via *remote* I/O.  I think the local file isn't necessarily a problem for everyone, because even though we use a centrally maintained bootstrap generator for TD, other environments might just distribute the bootstrap config with the clients.  But I do take your point that reading a file is still I/O and can be an issue in some environments.

FWIW, note that we *do* actually support a list of credentials for each server, and the client will stop at the first one that it supports, so there is a mechanism here for handling older clients.  But that doesn't help if there's a bug in parsing the config or in actually configuring the credentials, which is the sort of thing that we'd normally use an env var guard for.

You're also right that we don't yet support multiple servers, but I don't think that would actually cause any significant complications here, because there's a separate list of creds for each server.

Anyway, I've updated the language here.

Reply:
Oh, we do support a list; good. I couldn't understand how we had such an oversight. Good to see we didn't.

For a feature like this, it is useful for an implementation to fully-implement the feature and be tested before registering it as an possible option in the bootstrap. That way all the keys here are supported, or not. (Modulo Java which has tls but none of these.) If it is merged as a single PR, there's no problem, but if the work is split up...